### ISSUE TITLE
"Authentication Failure" when running SSHing into an Arch/Manjaro box

### ISSUE THREAD

**Issue by mike-hearn** [tailscale/tailscale#4924]
### What is the issue?

When trying to use Tailscale's SSH functionality into an Arch machine (in my case from MacOS -> Arch), it returns the error `Authentication Failure`

This issue was first flagged on the forum, and there's some additional information there - https://forum.tailscale.com/t/tailscale-ssh-authentication-failure-on-manjaro-arch/2564

### Steps to reproduce

1. Install Tailscale + SSH on an Arch box
2. Add that box to the Tailscale network, as well as whatever client machine you are using
3. SSH into the Arch box
4. Observe the failure message
5. Hold back your tears

### Are there any recent changes that introduced the issue?

It's the new beta SSH feature, so bugs are expected! I still love you, Tailscale.

### OS

Linux, macOS

### OS version

5.17.0-rc7-asahi-next-20220310-5-2-ARCH

### Tailscale version

1.26.1

### Bug report

BUG-b8f072d4dc1c38925c280914f94579d937dd0af19aacde6633fc20c7d5e0eed4-20220624193515Z-5ef9537512f1ef2f

**Comment by mike-hearn** [comment]
Here is some additional logging provided by sundbp on the Tailscale forum:

```Jun 22 23:16:04 thehost tailscaled[2623]: 100.78.255.16:53068->theuser@100.81.41.35:22: starting session: 20220622T221603-4880af8916
Jun 22 23:16:04 thehost tailscaled[2623]: ssh-session(20220622T221603-4880af8916): handling new SSH connection from theuser@gmail.com (100.78.255.16) to ssh-user "theuser"
Jun 22 23:16:04 thehost tailscaled[2623]: ssh-session(20220622T221603-4880af8916): access granted to theuser@gmail.com as ssh-user "theuser"
Jun 22 23:16:04 thehost tailscaled[2623]: ssh-session(20220622T221603-4880af8916): starting pty command: [/usr/bin/tailscaled be-child ssh --uid=1000 --gid=1000 --groups=1000,998,995,993,991,988,987,986,3,90,9>
Jun 22 23:16:04 thehost login[3240032]: pam_warn(remote:account): function=[pam_sm_acct_mgmt] flags=0 service=[remote] terminal=[/dev/pts/30] user=[theuser] ruser=[<unknown>] rhost=[100.78.255.16]
Jun 22 23:16:04 thehost login[3240032]: Authentication failure
```

**Comment by mfashby-tailscale** [comment]
I'm getting the same, but my work machine is also a manjaro laptop.
```
Jun 25 22:24:36 rpi3 tailscaled[532]: control: NetInfo: NetInfo{varies=true hairpin=false ipv6=false udp=true derp=#8 portmap= link=""}
Jun 25 22:24:36 rpi3 tailscaled[532]: magicsock: endpoints changed: 90.251.253.38:41643 (stun), 90.251.253.38:41641 (stun4localport), 192.168.1.66:41641 (local)
Jun 25 22:27:45 rpi3 tailscaled[532]: Accept: TCP{100.113.83.36:49378 > 100.69.31.118:22} 60 tcp ok
Jun 25 22:27:45 rpi3 tailscaled[532]: Accept: TCP{100.113.83.36:49378 > 100.69.31.118:22} 52 tcp non-syn
Jun 25 22:27:45 rpi3 tailscaled[532]: Accept: TCP{100.113.83.36:49378 > 100.69.31.118:22} 73 tcp non-syn
Jun 25 22:27:46 rpi3 tailscaled[532]: control: NetInfo: NetInfo{varies=false hairpin=false ipv6=false udp=true derp=#8 portmap= link=""}
Jun 25 22:27:46 rpi3 tailscaled[532]: magicsock: endpoints changed: 90.251.253.38:41643 (stun), 192.168.1.66:41641 (local)
Jun 25 22:27:46 rpi3 tailscaled[532]: 100.113.83.36:49378->martin@100.69.31.118:22: starting session: 20220625T212745-d35d5d0ffe
Jun 25 22:27:46 rpi3 tailscaled[532]: ssh-session(20220625T212745-d35d5d0ffe): handling new SSH connection from mfashby-tailscale@github (100.113.83.36) to ssh-user "martin"
Jun 25 22:27:46 rpi3 tailscaled[532]: ssh-session(20220625T212745-d35d5d0ffe): access granted to mfashby-tailscale@github as ssh-user "martin"
Jun 25 22:27:46 rpi3 tailscaled[532]: ssh-session(20220625T212745-d35d5d0ffe): starting pty command: [/usr/bin/tailscaled be-child ssh --uid=1000 --gid=1000 --groups=1000,3,
90,98,998,995,993,991,988,986,985,1001 --local-user=martin --remote-user=mfashby-tailscale@github --remote-ip=100.113.83.36 --has-tty=true --tty-name=pts/8 --shell --login-c
md=/usr/bin/login --cmd=/bin/bash -- -l]
Jun 25 22:27:46 rpi3 login[1232]: pam_warn(remote:account): function=[pam_sm_acct_mgmt] flags=0 service=[remote] terminal=[/dev/pts/8] user=[martin] ruser=[<unknown>] rhost=
[100.113.83.36]
Jun 25 22:27:46 rpi3 login[1232]: Authentication failure
Jun 25 22:27:49 rpi3 tailscaled[532]: ssh-session(20220625T212745-d35d5d0ffe): Wait: code=1
```

**Comment by SebDanielsson** [comment]
Same issue on Manjaro ARM@odroid-n2. Fully updated. Tailscale 1.26.0.

**Comment by mfrw** [comment]
Able to repro with tailscale version `1.26.1` on Arch.

**Comment by madkinsz** [comment]
Also able to reproduce on EndeavorOS. Here's some more debug information:

```
BUG-ada953a253e2071671c7aeb6b5bd35317d94f5734c8fea8b1addbae5f4f2d0ac-20220626222520Z-3756b65be27da3f5
```

```
Jun 26 17:25:14 mzvlka tailscaled[901]: ssh-session(20220626T222511-6ab9515051): Wait: code=1
Jun 26 17:25:11 mzvlka kernel: audit: type=1101 audit(1656282311.647:195): pid=33697 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:accounting grantors=? acct="mz" exe="/usr/bin/login" hostname=XXX addr=XXX terminal=/dev/pts/9 res=failed'
Jun 26 17:25:11 mzvlka login[33697]: Authentication failure
Jun 26 17:25:11 mzvlka audit[33697]: USER_ACCT pid=33697 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:accounting grantors=? acct="mz" exe="/usr/bin/login" hostname=XXX addr=XXX terminal=/dev/pts/9 res=failed'
Jun 26 17:25:11 mzvlka login[33697]: pam_warn(remote:account): function=[pam_sm_acct_mgmt] flags=0 service=[remote] terminal=[/dev/pts/9] user=[mz] ruser=[<unknown>] rhost=[XXX]
Jun 26 17:25:11 mzvlka tailscaled[901]: ssh-session(20220626T222511-6ab9515051): starting pty command: [/usr/bin/tailscaled be-child ssh --uid=1000 --gid=1000 --groups=1000,3,998,984,982 --local-user=mz --remote-user=madkinsz@github --remote-ip=XXX --has-tty=true --tty-name=pts/9 --shell --login-cmd=/usr/bin/login --cmd=/bin/bash -- -l]
Jun 26 17:25:11 mzvlka tailscaled[901]: ssh-session(20220626T222511-6ab9515051): access granted to madkinsz@github as ssh-user "mz"
Jun 26 17:25:11 mzvlka tailscaled[901]: ssh-session(20220626T222511-6ab9515051): handling new SSH connection from madkinsz@github (XXX ) to ssh-user "mz"
Jun 26 17:25:11 mzvlka tailscaled[901]: 100.115.94.95:50305->mz@XX: starting session: 20220626T222511-6ab9515051
Jun 26 17:25:11 mzvlka tailscaled[901]: Accept: TCP{XXX:50305 > XXX:22} 73 tcp non-syn
Jun 26 17:25:11 mzvlka tailscaled[901]: Accept: TCP{XXX:50305 > XXX:22} 52 tcp non-syn
Jun 26 17:25:11 mzvlka tailscaled[901]: Accept: TCP{XXX:50305 > XXX:22} 64 tcp ok
```

```
[mz@mzvlka ~]$ uname -r
5.18.7-arch1-1
```

```
[mz@mzvlka ~]$ tailscale --version
1.26.1
  tailscale commit: 5b81baa7d367d353707e87e0f9a8062cccc27c1b
  go version: go1.18.3
```

**Comment by eddiezane** [comment]
I got super curious and learned enough about PAM to figure this out.

The files in `/etc/pam.d/` should match the name of the command being called ("service" in PAM speak). I couldn't understand where the `remote` in `pam_warn(remote:account)` was coming from when it should have been `login` from `--login-cmd=/usr/bin/login`.

Checking on a Raspberry Pi running Debian I can see that `pam_unix(login:session)` is called successfully on `tailscale ssh`. 

Why is this showing up as the `login` I expect but Arch is calling `remote`?

After some digging it turns out that Debian's `login` is provided by the [shadow-utils](https://github.com/shadow-maint/shadow) package while Arch's is provided by [util-linux](https://github.com/util-linux/util-linux).

When Arch's `login` is called with `-h` it sets the PAM service name to `remote`. This is done by `tailscaled` [here](https://github.com/tailscale/tailscale/blob/bef6e2831a9b03e77ce1276ce94f8295061becc4/ssh/tailssh/incubator_linux.go#L178
).

https://github.com/shadow-maint/shadow/blob/274e786be970a659e6082b7733b451349335252c/src/login.c#L710

https://github.com/util-linux/util-linux/blob/dcbddb95364a9da6483c2b801f595497a6347996/login-utils/login.c#L890

Finding this in the manpage would have saved some time :sweat_smile:.

```
  -h
           Used by other servers (such as telnetd(8) to pass the name of the remote host to login so that it can be placed in utmp and wtmp. Only the
           superuser is allowed use this option.

           Note that the -h option has an impact on the PAM service name. The standard service name is login, but with the -h option, the name is remote. It
           is necessary to create proper PAM config files (for example, /etc/pam.d/login and /etc/pam.d/remote).
```

So the solution seems to be to `cp /etc/pam.d/login /etc/pam.d/remote`. This works but someone who knows more than me should confirm this is the right fix.

**Comment by bradfitz** [comment]
Debug trophy goes to @eddiezane. 🏆

**Comment by bradfitz** [comment]
FWIW, `man 1 login` on Debian/Ubuntu says only:

```
       -h
           Name of the remote host for this login.
```

As a quick workaround fix, we could not pass `-h` on Arch-based distros for now when the `/etc/pam.d/remote` file doesn't exist? It's not strictly required.

Does that seem okay to start?

**Comment by bradfitz** [comment]
If so, I sent https://github.com/tailscale/tailscale/pull/4963

**Comment by bradfitz** [comment]
Submitted. We should be able to cherry-pick that into the next point release too.

**Comment by raggi** [comment]
This also affected my Clearlinux instances. Clear is using empty-by-default `/etc`, so the heuristics would look different. i.e. pam.d exists in /usr/share/pam.d, as well as /etc/pam.d, with the prior being the commonly populated path. It does not include a remote configuration be default.

It is possible to simply grep login for util-linux to detect this login version, but I've not yet found a reliable way to determine the total set of pam configuration paths.

**Comment by DentonGentry** [comment]
We believe this is fixed by https://github.com/tailscale/tailscale/commit/029508d8bec8622979d08d8cbebaafec22b4dca9 in Tailscale 1.30, to be released next week. It can be tried now using an unstable release: https://pkgs.tailscale.com/unstable/

**Comment by DentonGentry** [comment]
Fixed in 1.30.
